homeassistant/components/file_upload/__init__.py

```diff
@@ -146,18 +151,32 @@ async def _upload_file(self, request: web.Request) -> web.Response:
         file_upload_data: FileUploadData = hass.data[DOMAIN]
         file_dir = file_upload_data.file_dir(file_id)
 
-        def _sync_work() -> None:
+        def _sync_create_dir() -> None:
             file_dir.mkdir()
 
-            # MyPy forgets about the isinstance check because we're in a function scope
-            assert isinstance(file_field, web.FileField)
+        await hass.async_add_executor_job(_sync_create_dir)
```

Comment:
Move the dir creation code to when you set file handle. That saves a trip to the executor.